java/com/google/re2j/RE2.java

```diff
@@ -213,25 +214,43 @@ int numberOfCapturingGroups() {
   // get() returns a machine to use for matching |this|.  It uses |this|'s
   // machine cache if possible, to avoid unnecessary allocation.
   Machine get() {
+    // Treiber stack (if reusing nodes) suffers from ABA problem on pop. Avoid by unlinking the
+    // entire stack, and stashing it in a pop-only stack guarded by a lock. This also reduces
+    // contention on the AtomicReference between putters and the getter.
```

Comment:
> reduces contention

How is that? Calls to get must incur a semaphore (exclusive lock).

Reply:
Having the two stacks and batch moves between them means that there's reduced CAS traffic on the AtomicReference, because getters either do a read (empty stack) or pop from their own stack, or import a whole batch, which amortizes the CAS on get.

Reply:
Right, but before reaching the read, they must first enter a synchronized block, whose semaphore requires an atomic decrement. How does the performance compare to a single stack?

Reply:
I can't use a single stack without allocating a new object per put(). That would make the whole system lock-free, but would also mean allocating on every call (once there's more than one concurrent call ever).

Or I could use a single stack and guard the pop section with a lock - but this would mean locking *and* atomic operations on every get() call. You would, however, be able to use a double-checked-locking approach to avoid synchronization if the pool is empty:

```java
    Machine head = machine.get();
    if (head != null) {
      // Lock necessary here because we're reusing nodes - otherwise suffer from ABA problem
      synchronized (this) {
        while (true) {
          head = machine.get();
          if (head == null) {
            break;
          }
          if (machine.compareAndSet(head, head.next)) {
            head.next = null;
            return head;
          }
          head = machine.get();
        }
      }
    }
```

That approach felt a bit more complicated and seemed like more of a departure from the existing code, but if you prefer it I can use it instead. Actual performance cost depends on target architecture, amount of contention, etc. The checked-in benchmarks for RE2 don't exercise threads, and even if there were concurrent benchmarks it'd be hard to generalize them across platforms.